rockpi4.mk

```diff
@@ -0,0 +1,231 @@
+# Partition no. 5 ends at 12288 + BR2_TARGET_ROOTFS_EXT2_SIZE (in kiB)
+.PHONY: boot-img
+boot-img: u-boot buildroot $(LINUX_PATH)/arch/arm64/boot/Image.gz
+	mkdir -p $(BINARIES_PATH)
+	rm -f $(BOOT_IMG)
+	truncate -s 128MiB $(BOOT_IMG)
+	parted -s $(BOOT_IMG) \
+		unit kiB \
+		mklabel gpt \
+		mkpart idbloader 32 4032 \
+		mkpart primary fat32 4032 4096 \
+		mkpart primary fat32 4096 8192 \
+		mkpart uboot 8192 12288 \
+		mkpart root fat32 12288 126976
+	sgdisk -u 5:17d61bff-8fdc-4089-b675-9be21b9f6ac7 $(BOOT_IMG)
+	dd if=$(UBOOT_PATH)/idbloader.img of=$(BOOT_IMG) bs=1kiB seek=32 conv=notrunc
+	dd if=$(UBOOT_PATH)/u-boot.itb of=$(BOOT_IMG) bs=1kiB seek=8192 conv=notrunc
+	e2mkdir $(ROOT_IMG):/boot
+	e2cp $(LINUX_PATH)/arch/arm64/boot/Image.gz $(ROOT_IMG):/boot
+	e2cp $(LINUX_PATH)/arch/arm64/boot/dts/rockchip/rk3399-rock-pi-4b.dtb $(ROOT_IMG):/boot
+	dd if=$(ROOT_IMG) of=$(BOOT_IMG) bs=1kiB seek=12288 conv=notrunc
+
+.PHONY: boot-img-clean
+boot-img-clean:
+	rm -f $(BOOT_IMG)
+
+clean: boot-img-clean
+
+################################################################################
+# rkdeveloptool
+################################################################################
+
+$(RKDEVELOPTOOL_PATH)/Makefile:
+	cd $(RKDEVELOPTOOL_PATH) && \
+		autoreconf -i && \
+		./configure CXXFLAGS=-Wno-format-truncation
+
+$(RKDEVELOPTOOL_BIN): $(RKDEVELOPTOOL_PATH)/Makefile
+	$(MAKE) -C $(RKDEVELOPTOOL_PATH)
+
+rkdeveloptool: $(RKDEVELOPTOOL_BIN)
+
+rkdeveloptool-clean:
+	$(MAKE) -C $(RKDEVELOPTOOL_PATH) clean
+
+rkdeveloptool-distclean:
+	$(MAKE) -C $(RKDEVELOPTOOL_PATH) clean
+
+clean: rkdeveloptool-clean
+
+$(LOADER_BIN):
+	cd $(BINARIES_PATH) && \
+		wget https://dl.radxa.com/rockpi/images/loader/$(notdir $(LOADER_BIN))
+
+################################################################################
+# Flash the image via USB onto the onboard eMMC
+################################################################################
+
+define flash-help
+	@echo
+	@echo "Please connect the board to the computer via a USB cable."
+	@echo "The cable must be connected to the upper USB 3 (blue) port."
+	@echo "Then press and hold the mask ROM button (first one on the left"
+	@echo "under the HDMI connector), apply power and release the button."
+	@echo "(More details at https://wiki.radxa.com/Rockpi4/dev/usb-install)"
+	@echo
+	@read -r -p "Press enter to continue, Ctrl-C to cancel:" dummy
+endef
+
+flash: $(BOOT_IMG) $(LOADER_BIN) $(RKDEVELOPTOOL_BIN)
+	$(call flash-help)
+	$(RKDEVELOPTOOL_BIN) db $(LOADER_BIN)
+	sleep 1
+	$(RKDEVELOPTOOL_BIN) wl 0 $(BOOT_IMG)
+
+nuke-emmc: $(LOADER_BIN) $(RKDEVELOPTOOL_BIN)
+	@echo
+	@echo "** WARNING: this command will make the onboard eMMC unbootable!"
+	@echo "It can be used to boot from the SD card again."
+	$(call flash-help)
+	dd if=/dev/zero of=$(BINARIES_PATH)/zero.img bs=1M count=64
+	$(RKDEVELOPTOOL_BIN) db $(LOADER_BIN)
+	sleep 1
+	$(RKDEVELOPTOOL_BIN) wl 0 $(BINARIES_PATH)/zero.img
```

Comment:
Is it possible to persuade `wl` to read from `stdin` instead? That way we wouldn't need a temporary file.

Reply:
Apparently not. The `rkdeveloptool` command needs a file name, it won't read from stdin. Worse: it needs a seekable file -- it does `lseek(... SEEK_END)` -- so bash process substitution `<()` won't work :-/

Reply:
OK, thanks for looking into it.